doc: openAIChat

MiriamScharnke opened the pull request:
Create functions directory in the doc folder and add openAIChat documentation.

ccreutzi committed:
Adding image support to ollamaChat

* Refactor `messageHistory` to be agnostic of the image encoding.
* Add backend encoding of images to `openAIChat`, `azureChat`, and `ollamaChat`.
* Add image test points to the test files.

Open question: Can we reliably detect which Ollama models support vision?

ccreutzi committed:
Add vision support to the doc files

ccreutzi committed:
Merge branch 'main' into ollama-images

ccreutzi committed:
Preinstall bakllava

Preinstall bakllava, to have a vision model

MiriamScharnke committed:
Text only edits to error message catalog (doc review)

ccreutzi committed:
Add doc warning about Ollama image input for text-only models

ccreutzi committed:
Update doc/Ollama.md

Co-authored-by: MiriamScharnke <mscharnk@mathworks.com>

ccreutzi committed:
WARNING -> TIP

MiriamScharnke committed:
Update +llms/+utils/errorMessageCatalog.m

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

MiriamScharnke committed:
Update +llms/+utils/errorMessageCatalog.m

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

ccreutzi committed:
Moved mlx into sub-directory, added Markdown export

Generating Markdown in pre-commit hook.

The hook needs to be activated in the local repository, by including it in `.git/hooks/pre-commit`.

ccreutzi committed:
Correct mlx path in texampleTests.m

ccreutzi committed:
Inserted text should have MATLAB®

Co-authored-by: MiriamScharnke <mscharnk@mathworks.com>

ccreutzi committed:
Also ignore data directory from moved mlx files

ccreutzi committed:
Add ® in generated files

Future versions will get that automatically, from commit 570aaa1044447ae81aab26f30462bd3ae70be627.

ccreutzi committed:
Merge branch 'generate-md-from-mlx' of github.com:matlab-deep-learning/llms-with-matlab into generate-md-from-mlx

MiriamScharnke committed:
Create functions directory in the doc folder and add openAIChat documentation.

ccreutzi committed:
Merge pull request #59 from matlab-deep-learning/ollama-images

Ollama images

ccreutzi commented on doc/functions/openAIChat.md:
````diff
@@ -0,0 +1,289 @@
+# Examples
+## Create OpenAI Chat
+```matlab
+modelName = "gpt-3.5-turbo";
````
We may not want to create new examples with 3.5. `"gpt-4o-mini"` is cheaper to use and generally considered on par or better.

MiriamScharnke replied:
Fixed

ccreutzi commented on doc/functions/openAIChat.md:
````diff
@@ -0,0 +1,289 @@
+chat = openAIChat(StreamFun=sf);
+generate(chat,"Why is a raven like a writing desk?")
+```
+# See Also
+-  [Create Simple Chat Bot](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/examples/CreateSimpleChatBot.md) 
+-  [Process Generated Text in Real Time Using ChatGPT in Streaming Mode](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/examples/ProcessGeneratedTextinRealTimebyUsingChatGPTinStreamingMode.md) 
+-  [Analyze Scientific Papers Using Function Calls](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/examples/AnalyzeScientificPapersUsingFunctionCalls.md) 
+-  [Analyze Sentiment in Text Using ChatGPT in JSON Mode](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/examples/AnalyzeSentimentinTextUsingChatGPTinJSONMode.md) 
````
```suggestion
# See Also
-  [Create Simple Chat Bot](../../examples/CreateSimpleChatBot.md) 
-  [Process Generated Text in Real Time Using ChatGPT in Streaming Mode](../../examples/ProcessGeneratedTextinRealTimebyUsingChatGPTinStreamingMode.md) 
-  [Analyze Scientific Papers Using Function Calls](../../examples/AnalyzeScientificPapersUsingFunctionCalls.md) 
-  [Analyze Sentiment in Text Using ChatGPT in JSON Mode](../../examples/AnalyzeSentimentinTextUsingChatGPTinJSONMode.md) 
```

ccreutzi reviewed (changes requested):
I wish I could mark individual comments as “request change” or “just an aside, feel free to ignore.”

I think the only thing where I have a flag is the `Tools` documentation. Let me know if the functionality there is completely opaque – we didn't come up with the terminology.

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -0,0 +1,289 @@
+OpenAI API key to access OpenAI APIs such as ChatGPT. 
+
+
+Instead of using the `APIKey` name\-value argument, you can also set the environment variable OPEN\_API\_KEY. For more information, see [https://github.com/matlab\-deep\-learning/llms\-with\-matlab/blob/main/doc/OpenAI.md](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/doc/OpenAI.md).
```
```suggestion
Instead of using the `APIKey` name\-value argument, you can also set the environment variable `OPEN_API_KEY`. For more information, see [the OpenAI overview documentation](../OpenAI.md).
```

MiriamScharnke replied:
For the API credentials: yes, I think that is a very good idea, just not on this page! I have created an issue to keep track of the suggestion, and will think about how to incorporate it - probably actually in the main readme, or in all three <third-party-API>.md pages.

MiriamScharnke committed:
Update doc/functions/openAIChat.md

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

MiriamScharnke committed:
Update doc/functions/openAIChat.md

I know why those happened, and can avoid them in the future!

In MLX files, you can create hyperlinks which link to lines in the same file. I tried to see whether this would translate into the markdown - turns out it hasn't. What it has done is taken those hyperlinks and just given them extra backticks with no functionality. So long as I stop trying to link between passages, this should not be an issue.

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

MiriamScharnke committed:
Update doc/functions/openAIChat.md

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

MiriamScharnke commented on doc/functions/openAIChat.md:
```diff
@@ -0,0 +1,289 @@
+
+To configure the format of the generated JSON file, describe the format using natural language and provide it to the model either in the system prompt or as a user message. For an example, see [Analyze Sentiment in Text Using ChatGPT in JSON Mode](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/examples/AnalyzeSentimentinTextUsingChatGPTinJSONMode.md). 
+
+# Properties
```
This is an on-going discussion in doc! 

The problem is that the name-value arguments and the properties overlap, but are not fully identical, nor is one a subset of the other. So the question becomes, how best to communicate this clearly? We want people to be able to tell easily what they can set via NV argument, and we also want people to tell easily what properties this object has.

A somewhat out-of-model option would be to have: 
Name-value arguments that do not correspond to properties
Properties settable at construction
Properties not settable at construction

This would have the advantage that no information would need to be repeated. For clarity, I could also add "chat = openAIChat(___,Property=value)" to the supported syntaxes. The disadvantage would be that this is out of model for the documentation. I'm not opposed to this, but would like to have a conversation about this in a team meeting. (I would also have to come up with better section titles - the above are just to illustrate the concept).

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -0,0 +1,289 @@
+
+## Description
+
+Connect to the OpenAI™ Chat Completion API to generate text using large language models developed by OpenAI.
```
Now we have the ™ twice on this page.
```suggestion
Connect to the OpenAI Chat Completion API to generate text using large language models developed by OpenAI.
```

ccreutzi committed:
Merge pull request #62 from matlab-deep-learning/generate-md-from-mlx

Moved mlx into sub-directory, added Markdown export

MiriamScharnke committed:
Apply suggestions from code review

Co-authored-by: Deborah Ferreira <dferreir@mathworks.com>

MiriamScharnke commented on doc/functions/openAIChat.md:
```diff
@@ -0,0 +1,289 @@
+`1` (default) | numeric scalar between `0` and `1`
+
+
+Top probability mass for controlling the diversity of the generated output. Higher top probability mass corresponds to higher diversity.
```
How about "Top probability mass for controlling the diversity of the generated output using top-p sampling. Higher top probability mass corresponds to higher diversity."? This does not mention nucleus sampling, but top-p sampling should be googleable enough!

MiriamScharnke committed:
Update doc/functions/openAIChat.md

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

MiriamScharnke committed:
Update doc/functions/openAIChat.md

Co-authored-by: Christopher Creutzig <89011131+ccreutzi@users.noreply.github.com>

MiriamScharnke committed:
Update openAIChat documentation.

MiriamScharnke committed:
Merge branch 'temp-change' into doc-openaichat

MiriamScharnke committed:
Merge pull request #60 from matlab-deep-learning/error-message-text-only-edits

Text only edits to error message catalog (doc review)

MiriamScharnke committed:
Fix OpenAI trademarks

MiriamScharnke committed:
Delete doc/functions/openAIChat.md

This file was not meant for this change.

MiriamScharnke committed:
Merge pull request #66 from matlab-deep-learning/fix-openai-trademark

Fix openai trademark

MiriamScharnke committed:
Technical feedback and cosmetic changes for openAIChat reference page

MiriamScharnke committed:
Technical feedback and cosmetic changes for openAIChat reference page

MiriamScharnke commented on doc/functions/openAIChat.md:
```diff
@@ -0,0 +1,289 @@
+`openAIFunction` object | array of `openAIFunction` objects
+
+
+Custom functions used by the model to process its input and output.
```
Fixed

MiriamScharnke committed:
Cosmetic changes for openAIChat reference page

MiriamScharnke committed:
Cosmetic changes for openAIChat reference page

MiriamScharnke committed:
Fix trademark on openAIChat reference page

MiriamScharnke committed:
Fix trademark on openAIChat reference page

ccreutzi committed:
Allow (long) char vectors for StopSequences

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -36,60 +37,78 @@ To connect to the OpenAI API, you need a valid API key. For information on how t
 
+
+`chat = openAIChat(___,PropertyName=PropertyValue)` specifies properties that are settable at construction using one or more name\-value arguments.
+
```
Listing both seems potentially confusing.

MiriamScharnke replied:
Fixed!

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -36,60 +37,78 @@ To connect to the OpenAI API, you need a valid API key. For information on how t
+
+### `Tools` – OpenAI functions to use during output generation 
+
+`openAIFunction` object | array of `openAIFunction` objects
```
I thought doc agreed that scalars are arrays?

MiriamScharnke replied:
I agree this is technically redundant, but still think it's helpful (because otherwise users may think they have to put their single function object inside square brackets, for example).

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -36,60 +37,78 @@ To connect to the OpenAI API, you need a valid API key. For information on how t
+For an example, see [Analyze Scientific Papers Using ChatGPT Function Calls](../../examples/AnalyzeScientificPapersUsingFunctionCalls.md). 
 
-### `ModelName` \- Model name
+## Properties Settable at Construction
```
`Tools` is such a property, I think?

(Looks like I cannot give a proposal to submit because of the deleted line in between. I recommend moving “Tools” down into this section.)

MiriamScharnke replied:
I don't think Tools is a property (so far as the user is concerned), only FunctionNames. You can't access "chat.Tools", for example ("No public property 'Tools' for class 'openAIChat'.").

ccreutzi replied:
Sorry, you're right. The `Tools` property is not exposed to the user, for whatever reason.

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -36,60 +37,78 @@ To connect to the OpenAI API, you need a valid API key. For information on how t
-### `StopSequences` \- Stop sequences
+### `StopSequences` – Stop sequences
 
 `""` (default) | string array with between `1` and `4` elements
```
Why do we want the numbers in monospace? Also, the default is an empty string array, not the one-element string with no characters.

```suggestion
`[]` (default) | string array with between zero and four elements
```

MiriamScharnke replied:
Fixed "" -> [] and 1 -> 0. The numbers are in monospace for consistency - if the string array had x elements, then x would be in monospace, too!

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -36,60 +37,78 @@ To connect to the OpenAI API, you need a valid API key. For information on how t
-Instead of using the `APIKey` name\-value argument, you can also set the environment variable OPEN\_API\_KEY. For more information, see [https://github.com/matlab\-deep\-learning/llms\-with\-matlab/blob/main/doc/OpenAI.md](https://github.com/matlab-deep-learning/llms-with-matlab/blob/main/doc/OpenAI.md).
+Instead of using the `APIKey` name\-value argument, you can also set the environment variable OPEN\_API\_KEY. For more information, see [OpenAI API](../OpenAI.md).
+
+---
```
These lines look heavier in the preview than the section break for “Properties Settable at Construction” below.

ccreutzi commented on doc/functions/openAIChat.md:
```diff
@@ -263,27 +226,28 @@ This property is read\-only.
 sf = @(x) fprintf("%s",x);
 chat = openAIChat(StreamFun=sf);
-generate(chat,"Why is a raven like a writing desk?")
+generate(chat,"Why is a raven like a writing desk?");
```
Generating text and discarding the output without saving it anywhere seems a bit pointless. Can we remove the semicolon?

MiriamScharnke replied:
We can; do you want me to actually generate some text and also print the output?

ccreutzi replied:
Possibly? Maybe faked or with an elision, so it doesn't get too long?

MiriamScharnke committed:
Cosmetic fixes

MiriamScharnke committed:
Cosmetic fixes

ccreutzi committed:
Merge pull request #67 from matlab-deep-learning/allow-char-StopSequences

Allow (long) char vectors for StopSequences

ccreutzi committed:
Replace bakllava by moondream

moondream is a much smaller vision model and perfectly suitable for the tests we want to run.

ccreutzi committed:
Create CODEOWNERS

ccreutzi committed:
Merge pull request #68 from matlab-deep-learning/replace-bakllava

Replace bakllava by moondream

ccreutzi committed:
Merge pull request #69 from matlab-deep-learning/codeowners

Create CODEOWNERS

ccreutzi committed:
Trace/replay `llms.internal.sendRequest`

To decouple `texampleTests.m` from availability and speed of external servers,
record calls to `llms.internal.sendRequest` (on dev machine) and replay
(during most test runs, including CI). See tests/recording/README.md for
instructions.

ccreutzi committed:
Avoid bogus access to `json.choices.delta.content`

The streaming interface sometimes responds with something like `json.choices = {"index":0,"delta":{},"logprobs":[],"finish_reason":"length"}`. Stop blindly assuming `json.choices.delta.content` exists.

ccreutzi committed:
Merge pull request #71 from matlab-deep-learning/avoid-content-error

Avoid bogus access to `json.choices.delta.content`

ccreutzi committed:
Merge pull request #70 from matlab-deep-learning/test-doubles

Trace/replay `llms.internal.sendRequest`

MiriamScharnke committed:
Cosmetic fixes

MiriamScharnke committed:
Cosmetic fixes

MiriamScharnke committed:
Cosmetic fixes

MiriamScharnke committed:
Merge branch 'doc-openaichat' of github.com:matlab-deep-learning/llms-with-matlab into doc-openaichat

MiriamScharnke committed:
Merge remote-tracking branch 'origin' into doc-openaichat
